jax/_src/lax/lax.py

```diff
@@ -4325,11 +4341,49 @@ def _iota_abstract_eval(*, dtype, shape, dimension):
 iota_p.def_impl(partial(xla.apply_primitive, iota_p))
 iota_p.def_abstract_eval(_iota_abstract_eval)
 
-def _iota_lower(ctx, *, dtype, shape, dimension):
-  del dtype, shape
+def _iota_staging_rule(trace, *dyn_shape, dtype, shape, dimension):
+  params = dict(dtype=dtype, shape=shape, dimension=dimension)
+  if not dyn_shape:
+    return trace.default_process_primitive(iota_p, (), params)
+  source_info = source_info_util.current()
+  ds = iter(dyn_shape)
+  out_shape_for_tracer = [next(ds) if d is None else d for d in shape]
+  aval = core.DShapedArray(tuple(out_shape_for_tracer), dtype, False)
+  out_tracer = pe.DynamicJaxprTracer(trace, aval, source_info)
+  invars = [trace.getvar(d) for d in dyn_shape]
+  eqn = pe.new_jaxpr_eqn(invars, [trace.makevar(out_tracer)],
+                         iota_p, params, core.no_effects, source_info)
+  trace.frame.eqns.append(eqn)
+  return out_tracer
+pe.custom_staging_rules[iota_p] = _iota_staging_rule
+
+def _iota_typecheck_rule(*dyn_shape, dtype, shape, dimension):
+  if not dyn_shape:
+    out_aval, effects = iota_p.abstract_eval(
+        dtype=dtype, shape=shape, dimension=dimension)
+    return [out_aval], effects
+  else:
+    dyn_shape_ = iter(dyn_shape)
+    out_shape = [next(dyn_shape_) if d is None else d for d in shape]
+    out_shape = [x.val if type(x) is core.Literal else x for x in out_shape]
+    out_aval = core.DShapedArray(tuple(out_shape), dtype, False)
+    return [out_aval], core.no_effects
+core.custom_typechecks[iota_p] = _iota_typecheck_rule
+
+def _iota_lower(ctx, *dyn_shape, dtype, shape, dimension):
+  del dtype
   aval_out, = ctx.avals_out
-  return mhlo.IotaOp(mlir.aval_to_ir_type(aval_out),
-                     mlir.i64_attr(dimension)).results
+  if dyn_shape:
+    dyn_shape_ = iter(dyn_shape)
+    shape = [next(dyn_shape_) if d is None else d for d in shape]
+    return mhlo.DynamicIotaOp(
```

Comment:
Because MHLO has distinct `IotaOp` and `DynamicIotaOp` operations, we need to switch on which to call here in the MHLO lowering rule.